url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/engine/impl/ProblemToSTS.java

```diff
@@ -0,0 +1,759 @@
+			Adaptation adaptation = corr.getAdaptation();
+			STS adSTS = this.adaptationSTSs.get(adaptation);
+
+			State initial = adSTS.getInitialState();
+			int prevIndex = this.getIndexOfPreviousAdaptation(index);
+			for (Transition initTrans : adSTS.getTransitionsFromState(initial)) {
+				for (int i = prevIndex; i <= index; i++) {
+					Action act = initTrans.getAction();
+					if (!act.getName().contains("Resume")) {
+						sem.addAction(act);
+						Transition newTrans = new DefaultTransition(stateList.get(i), 
+								new StateFormula(), act, stateList.get(index));
+						sem.addTransition(newTrans);      
+					}
+				}
+			}
+			Action action = adSTS.getAction("Resume"+index);
+			sem.addAction(action);
+//			Transition resTrans1 = new DefaultTransition(initState,
+//					new StateFormula(), action, initState);
+//			sem.addTransition(resTrans1);
+			Transition resTrans = new DefaultTransition(stateList.get(index),
+					new StateFormula(), action, initState);
+			sem.addTransition(resTrans);
+			index++;
+		}
+		return sem;
+	}
+
+	private STS transformDomainObject(DomainObject obj) {
+		init();
+		State initialState = null;
+		for (ObjectState state : obj.getStates()) {
+			State newState = createState(state.getName());
+			State initial = obj.getInitialState();
+			if (initial != null && initial.equals(state)) {
+				initialState = newState;
+			}
+		}
+		
+		Map<ObjectEvent, Map<Action, Activity>> event2action = 
+						getRelatedActions(obj);
+
+		for (ObjectTransition trans : obj.getTransitions()) {
+			ObjectEvent event = trans.getObjectEvent();
+			Map<Action, Activity> entries = event2action.get(event);
+			if (entries == null || entries.isEmpty()) {
+				continue;
+			}
+			
+			for (Action action : entries.keySet()) {
+				State startState = null;
+				State endState = null;
+				
+				for (State state : states) {
+					if (state.getName().equals(trans.getStartState().getName())) {
+						startState = state;
+					}
+					if (state.getName().equals(trans.getEndState().getName())) {
+						endState = state;
+					}
+				}
+				if (startState != null && endState != null) { 
+					createTransition(startState, new StateFormula(), 
+						action, endState);
+				}
+			}
+		}		
+		return new DefaultSTS(states, initialState, inputActions, 
+				outputActions, transitions);
+	}
+	
+	
+	private Map<ObjectEvent, Map<Action, Activity>> getRelatedActions(DomainObject object) {
+		
+		logger.trace("Collecting actions related to domain object " + object.getName());
+		Map<ObjectEvent, Map<Action, Activity>> relatedActions =
+			new HashMap<ObjectEvent, Map<Action, Activity>>();
+		
+		for (ObjectEvent event : object.getEvents()) {
+			
+			Map<Action, Activity> eventActions = new HashMap<Action, Activity>();
+			EventLiteral evLit = new EventLiteral(object, event);
+			
+			Map<Action, Activity> actionTable = model2sts.getActionTable();
+			
+			for (Action action : actionTable.keySet()) {
+				Activity activity = actionTable.get(action);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.